database/src/service/read.rs

```diff
@@ -211,6 +226,71 @@ fn map_request(
     /* TODO: post-request handling, run some code for each request? */
 }
 
+//---------------------------------------------------------------------------------------------------- Thread Local
+/// Q: Why does this exist?
+///
+/// A1: `heed`'s transactions and tables are not `Sync`, so we cannot use
+/// them with rayon, however, we set a feature such that they are `Send`.
+///
+/// A2: When sending to rayon, we want to ensure each read transaction
+/// is only being used by 1 thread only to scale reads
+///
+/// <https://github.com/Cuprate/cuprate/pull/113#discussion_r1576762346>
+#[inline]
+fn thread_local<T: Send>(env: &impl Env) -> ThreadLocal<T> {
+    ThreadLocal::with_capacity(env.config().reader_threads.as_threads().get())
+}
+
+/// Depending on the backend, the table type will need `UnsafeSendable`.
+///
+/// This macro branches on the backend and uses the appropriate type.
+/// See the below `SAFETY` comment for more info.
+///
+/// # Notes
+/// This macro will early return the scope it is called from
+/// with an `Err(RuntimeError)` if there is a database error.
+macro_rules! get_tables {
+    ($env_inner:ident, $tx_ro:ident, $tables:ident) => {{
+        cfg_if::cfg_if! {
+            if #[cfg(all(feature = "redb", not(feature = "heed")))] {
+                $tables.get_or_try(|| $env_inner.open_tables($tx_ro))?
+            } else {
+                $tables.get_or_try(|| {
+                    #[allow(clippy::significant_drop_in_scrutinee)]
+                    match $env_inner.open_tables($tx_ro) {
+                        // SAFETY:
+                        // `heed`'s transactions are `Send` but `HeedTableRo` contains a `&`
+                        // to the transaction, as such, if `Send` were implemented on `HeedTableRo`
+                        // then 1 transaction could be used to open multiple tables, then sent to
+                        // other threads - this would be a soundness hole against `Sync`.
+                        //
+                        // `&T` is only `Send` if `T: Sync`.
+                        //
+                        // `heed::RoTxn: !Sync`, therefore our table
+                        // holding `&heed::RoTxn` must NOT be `Send`.
+                        //
+                        // <https://doc.rust-lang.org/std/marker/trait.Sync.html>
+                        // <https://doc.rust-lang.org/nomicon/send-and-sync.html>
+                        //
+                        // Instead, we are unsafely using `UnsafeSendable` in `service`'s reader
+                        // thread-pool as we are pairing our usage with `ThreadLocal` - only 1 thread
+                        // will ever access a transaction at a time. This is an INVARIANT.
```

Comment:
Did you mean safely here?

```suggestion
                        // Instead, we are safely using `UnsafeSendable` in `service`'s reader
                        // thread-pool as we are pairing our usage with `ThreadLocal` - only 1 thread
                        // will ever access a transaction at a time. This is an INVARIANT.
```